src/Hamiltonians.jl

```diff
@@ -1618,4 +1654,426 @@ function m1p(bsadd::BStringAdd,chosen::Int,j::Int)
+        # println("Hop in B, chosen = $chosen") # debug
+        return BoseFS2C{NA,NB,M,AA,AB}(add.bsa,naddress_from_bsb), elem
+    else
+        chosen -= (nhops_a+nhops_b)
+        sa = numberoccupiedsites(add.bsa)
+        sb = numberoccupiedsites(add.bsb)
+        # println("Hops across A and B: $(sa*(ham.m-1)*sb)")
+        new_bsa, new_bsb, onproduct_a, onproduct_b = hopacross2adds(add.bsa, add.bsb, chosen)
+        new_add = BoseFS2C{NA,NB,M,AA,AB}(new_bsa,new_bsb)
+        # println("Hop A to B, chosen = $chosen") # debug
+        # return new_add, elem
+        elem = ham.v/ham.ha.m*sqrt(onproduct_a)*sqrt(onproduct_b)
+        new_add = BoseFS2C{NA,NB,M,AA,AB}(new_bsa,new_bsb)
+        return new_add, elem
+    end
+    # return new address and matrix element
+end
+
+# hopacross2adds needed for computing hops across two components
+@inline function hopacross2adds(add_a::BoseFS{NA,M,AA}, add_b::BoseFS{NB,M,AB}, chosen::Integer) where {NA,NB,M,AA,AB}
+    sa = numberoccupiedsites(add_a)
+    sb = numberoccupiedsites(add_b)
+    onrep_a = BitStringAddresses.s_onr(add_a)
+    onrep_b = BitStringAddresses.s_onr(add_b)
+    # b†_s b_q a†_p a_r
+    s = p = q = r = 0
+    onproduct_a = 1
+    onproduct_b = 1
+    hole_a, remainder = fldmod1(chosen, (M-1)*sb) # hole_a: position for hole_a
+    p, hole_b = fldmod1(remainder, sb) # hole_b: position for hole_b
+    # annihilate an A boson:
+    for (i, occ) in enumerate(onrep_a)
+      if occ > 0
+        hole_a -= 1 # searching for the position for hole_a
+        if hole_a == 0 # found the hole_a here
+          onproduct_a *= occ # record the normalisation factor before annihilate
+          onrep_a = @set onrep_a[i] = occ-1 # annihilate an A boson: a_r
+          r = i # remember where we make the hole
+          break # should break out of the for loop
+        end
+      end
+    end
+    if p ≥ r
+        p += 1 # to skip the hole_a
+    end
+    # create an A boson:
+    ΔP = p-r # change in momentun
+    p = mod1(p, M) # enforce periodic boundary condition
+    onrep_a = @set onrep_a[p] += 1 # create an A boson: a†_p
+    onproduct_a *= onrep_a[p] # record the normalisation factor after creation
+    # annihilate a B boson:
+    for (i, occ) in enumerate(onrep_b)
+      if occ > 0
+        hole_b -= 1 # searching for the position for hole_b
+        if hole_b == 0 # found the hole_b here
+          onproduct_b *= occ # record the normalisation factor before annihilate
+          onrep_b = @set onrep_b[i] = occ-1 # annihilate a B boson: b_q
+          q = i # remember where we make the holes
+          break # should break out of the for loop
+        end
+      end
+    end
+    s = mod1(q-ΔP, M) # compute s with periodic boundary condition
+    # create a B boson:
+    onrep_b = @set onrep_b[s] += 1 # create a B boson: b†_s
+    onproduct_b *= onrep_b[s] # record the normalisation factor after creation
+    if mod(q+r,M)-mod(s+p,M) != 0 # sanity check for momentum conservation
```

Comment:
I suggest to remove (or comment out) this if condition (three lines) from the production code. After you have checked that your code produces the correct Hamiltonian matrix, there is no need to check at run time every time you generate an off-diagonal matrix element. This will just slow down your code.

Reply:
Commented out.